File: tools/client/model/vehicle.py
from google.cloud.storage import Bucket
from client.model.entity import Entity
from client.utils.filters import apply_filters
from google.cloud.firestore import DocumentReference
from client.model.collection_unit import CollectionUnit


class Snapshot(Entity):

    content: str

    def __init__(self, data):
        super().__init__(data)
        self.content = data['content']


class Video(Entity):

    videoId: str

    def __init__(self, data, storage: Bucket):
        super().__init__(data)
        self.videoId = data['videoId']
        self.storage = storage

    async def getVideoUrl(self):
        return self.storage.blob(
            f"videos/{self.videoId}.mp4").generate_signed_url()


class Deformation(Entity):

    content: str

    def __init__(self, data):
        super().__init__(data)
        self.content = data['content']


class VehicleInfo:

    carId: str

    def __init__(self, data):
        self.carId = data['carId']

    async def load(self):
        """
        TODO: Load real vehicle info from DB
        """
        pass

    @property
    def modelName(self):
        """
        TODO: Load real vehicle info from DB
        """
        return "Elantra"


class VehicleMetrics:

    def __init__(self, data):
        self.batteryLevel = data['batteryLevel']
        self.maximumRange = data['maximumRange']
        self.distanceDriven = data['distanceDriven']


class Vehicle(Entity):

    id: str
    createdAt: int
    serialCode: str
    info: VehicleInfo
    metrics: VehicleMetrics
    snapshots: CollectionUnit[Snapshot]
    videos: CollectionUnit[Video]
    deformations: CollectionUnit[Deformation]

    def __init__(self, data):
        super().__init__(data)
        self.id = data['id']
        self.createdAt = data['createdAt']
        self.serialCode = data['serialCode']
        self.info = VehicleInfo(data['info'])
        self.metrics = VehicleMetrics(data['metrics'])
        self.snapshots = data['snapshots']
        self.videos = data['videos']
        self.deformations = data['deformations']

    @staticmethod
    async def from_ref(ref: DocumentReference, storage: Bucket):
        document = await ref.get()
        if not document.exists:
            raise Exception("[FirebaseException] Document doesn't exist")
        data = document.to_dict()
        return Vehicle({
            'id':
            data['id'],
            'createdAt':
            data['createdAt'],
            'serialCode':
            data['serialCode'],
            'info':
            data['info'],
            'metrics':
            data['metrics'],
            'snapshots':
            CollectionUnit(ref.collection('snapshot'),
                           lambda data: Snapshot(data)),
            'videos':
            CollectionUnit(ref.collection('video'),
                           lambda data: Video(data, storage)),
            'deformations':
            CollectionUnit(ref.collection('deformation'),
                           lambda data: Deformation(data))
        })

    async def listSnapshots(self, options):
        return self.snapshots.filter(
            lambda collection: apply_filters(collection, options))

    async def listVideos(self, options):
        return self.videos.filter(
            lambda collection: apply_filters(collection, options))

    async def listDeformations(self, options):
        return self.deformations.filter(
            lambda collection: apply_filters(collection, options))

    async def getSnapshot(self, id):
        snapshots = await self.snapshots.filter(
            lambda collection: collection.where('id', '==', id).limit(1))
        if not snapshots:
            raise Exception(
                f"[FirebaseException] Snapshot with id {id} doesn't exist")
        return snapshots[0]

    async def getVideo(self, id):
        videos = await self.videos.filter(
            lambda collection: collection.where('id', '==', id).limit(1))
        if not videos:
            raise Exception(
                f"[FirebaseException] Video with id {id} doesn't exist")
        return videos[0]

    async def getDeformation(self, id):
        deformations = await self.deformations.filter(
            lambda collection: collection.where('id', '==', id).limit(1))
        if not deformations:
            raise Exception(
                f"[FirebaseException] Deformation with id {id} doesn't exist")
        return deformations[0]

File: libs/firebase/src/lib/model/vehicle.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { CollectionUnit } from '../collection';
import { applyFilters, FilterOptions } from '../filter-options';
import { DocumentReference, FirebaseStorage } from '../firebase-types';
import { Entity } from './base';

export class Snapshot extends Entity {
  content: string;

  constructor(data: Snapshot) {
    super(data);
    this.content = data.content;
  }
}

interface IVideo extends Entity {
  /**
   * @description
   * 이 id로 firebase storage에 저장
   */
  videoId: string;
}

export class Video extends Entity implements IVideo {
  videoId: string;

  constructor(data: IVideo, private readonly storage: FirebaseStorage) {
    super(data);
    this.videoId = data.videoId;
  }

  /**
   * @todo Handle other video formats
   */
  async getVideoUrl(): Promise<string> {
    return this.storage.ref(`videos/${this.videoId}.mp4`).getDownloadURL();
  }
}

export class Deformation extends Entity {
  content: string;

  constructor(data: Deformation) {
    super(data);
    this.content = data.content;
  }
}

interface IVehicleInfo {
  /**
   * @description
   * 차량 고유 번호
   */
  carId: string;
}

class VehicleInfo implements IVehicleInfo {
  carId: string;

  constructor(data: IVehicleInfo) {
    this.carId = data.carId;
  }

  /**
   * @todo Load real model from server or API
   */
  async load() {
    return;
  }

  /**
   * @todo Return real model name
   */
  get modelName(): string {
    return 'Elantra';
  }
}

interface VehicleMetrics {
  /**
   * @description
   * 0 ~ 1
   */
  batteryLevel: number;
  /**
   * @description
   * Unit in km
   */
  maximumRange: number;
  distanceDriven: number;
}

export interface IVehicle extends Entity {
  serialCode: string;
  info: VehicleInfo;
  metrics: VehicleMetrics;
  snapshots: CollectionUnit<Snapshot>;
  videos: CollectionUnit<Video>;
  deformations: CollectionUnit<Deformation>;
}

export class Vehicle implements IVehicle {
  id: string;
  createdAt: number;
  serialCode: string;
  info: VehicleInfo;
  metrics: VehicleMetrics;
  snapshots: CollectionUnit<Snapshot>;
  videos: CollectionUnit<Video>;
  deformations: CollectionUnit<Deformation>;

  constructor(data: IVehicle) {
    this.id = data.id;
    this.createdAt = data.createdAt;
    this.serialCode = data.serialCode;
    this.info = new VehicleInfo(data.info);
    this.metrics = data.metrics;
    this.snapshots = data.snapshots;
    this.videos = data.videos;
    this.deformations = data.deformations;
  }

  static async fromRef(ref: DocumentReference, storage: FirebaseStorage): Promise<Vehicle> {
    const document = await ref.get();
    if (!document.exists) {
      throw new Error("[FirebaseException] Document doesn't exist");
    }
    const { serialCode, info, id, createdAt, metrics } = document.data() as IVehicle;
    return new Vehicle({
      id,
      createdAt,
      serialCode,
      info,
      metrics,
      snapshots: new CollectionUnit(ref.collection('snapshot') as any, (data) => new Snapshot(data)),
      videos: new CollectionUnit(ref.collection('video') as any, (data) => new Video(data, storage)),
      deformations: new CollectionUnit(ref.collection('deformation') as any, (data) => new Deformation(data)),
    });
  }

  async listSnapshots(options: FilterOptions): Promise<Snapshot[]> {
    return this.snapshots.filter((collection) => applyFilters(collection, options));
  }

  async listVideos(options: FilterOptions): Promise<Video[]> {
    return this.videos.filter((collection) => applyFilters(collection, options));
  }

  async listDeformations(options: FilterOptions): Promise<Deformation[]> {
    return this.deformations.filter((collection) => applyFilters(collection, options));
  }

  async getSnapshot(id: string): Promise<Snapshot> {
    const snapshots = await this.snapshots.filter((collection) => collection.where('id', '==', id).limit(1));
    if (snapshots.length === 0) throw new Error(`[FirebaseException] Snapshot with id ${id} doesn't exist`);
    return snapshots[0];
  }

  async getVideo(id: string): Promise<Video> {
    const videos = await this.videos.filter((collection) => collection.where('id', '==', id).limit(1));
    if (videos.length === 0) throw new Error(`[FirebaseException] Video with id ${id} doesn't exist`);
    return videos[0];
  }

  async getDeformation(id: string): Promise<Deformation> {
    const deformations = await this.deformations.filter((collection) => collection.where('id', '==', id).limit(1));
    if (deformations.length === 0) throw new Error(`[FirebaseException] Deformation with id ${id} doesn't exist`);
    return deformations[0];
  }
}
